### PR TITLE
Implement sleeper agent hidden trigger backdoor attack in PyTorch

### LINKED ISSUE #1392: Implement sleeper agent
**Is your feature request related to a problem? Please describe.**
Sleeper Agent: Scalable Hidden Trigger Backdoors for Neural Networks Trained from Scratch
https://arxiv.org/pdf/2106.08970.pdf 

**Describe the solution you'd like**
- jupyter notebooks

### PR DESCRIPTION
**Pull request by monshri**
# Description

- [x] Implement sleeper agent hidden trigger backdoor attack in PyTorch Framework.
- [x] Replacing double underscore functions in GradientMatchingAttack class to single underscore for supporting inheritance.

Fixes #1392

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+| Paper link: https://arxiv.org/abs/2106.08970
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Tuple, TYPE_CHECKING, List
+import random
+from tqdm.auto import trange
+import numpy as np
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:29]:
```suggestion
import logging
from typing import Tuple, TYPE_CHECKING, List
import random

import numpy as np
from tqdm.auto import trange
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    # pylint: disable=C0412
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class SleeperAgentAttack(GradientMatchingAttack):
+    """
+    Implementation of Sleeper Agent Attack
+    | Paper link: https://arxiv.org/pdf/2106.08970.pdf
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:44]:
```suggestion
    Implementation of Sleeper Agent Attack

    | Paper link: https://arxiv.org/pdf/2106.08970.pdf
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        batch_size: int = 128,
+        clip_values: Tuple[float, float] = (0, 1.0),
+        verbose: int = 1,
+        indices_target=None,
+        patching_strategy="random",
+        selection_strategy="random",
+        retraining_factor=1,
+        model_retrain=False,
+        model_retraining_epoch=1,
+        patch=None,
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:64]: Please add typing.

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        if isinstance(self.substitute_classifier, TensorFlowV2Classifier):
+            poisoner = self._poison__tensorflow
+            finish_poisoning = self._finish_poison_tensorflow
+            initializer = self._initialize_poison_tensorflow
+        elif isinstance(self.substitute_classifier, PyTorchClassifier):
+            poisoner = self._poison__pytorch
+            finish_poisoning = self._finish_poison_pytorch
+            initializer = self._initialize_poison_pytorch
+        else:
+            raise NotImplementedError("SleeperAgentAttack is currently implemented only for Tensorflow V2 and Pytorch.")
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:136]:
```suggestion
            raise NotImplementedError("SleeperAgentAttack is currently implemented only for TensorFlow V2 and PyTorch.")
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+    def get_poison_indices(self):
+        """
+        :return: indices of best poison index
+        """
+        return self.indices_poison
+
+    def model_retraining(self, poisoned_samples):
+        """
+        Applies retraining to substitute model
+        :param poisoned_samples: poisoned array.
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:202]:
```suggestion
        Applies retraining to substitute model

        :param poisoned_samples: poisoned array.
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+        from art.estimators.classification.pytorch import PyTorchClassifier
+
+        (x_train, y_train), (x_test, y_test), min_, max_ = load_cifar10()
+        mean = np.mean(x_train, axis=(0, 1, 2, 3))
+        std = np.std(x_train, axis=(0, 1, 2, 3))
+        x_train = (x_train - mean) / (std + 1e-7)
+        x_test = (x_test - mean) / (std + 1e-7)
+        min_ = (min_ - mean) / (std + 1e-7)
+        max_ = (max_ - mean) / (std + 1e-7)
+        x_train = np.transpose(x_train, [0, 3, 1, 2])
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:214]: This method is specific for CIFAR10. How about if we ask for the training data as input?

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+        """
+        Creates new model
+        :param x_train: Samples of train data.
+        :param y_train: Labels of train data.
+        :param x_test: Samples of test data.
+        :param y_test: Labels of test data.
+        :param num_classes: Number of classes of labels in train data.
+        :param batch_size: The size of batch used for training.
+        :param epochs: The number of epochs for which training need to be applied.
+        :return model, loss_fn, optimizer - trained model, loss function used to train the model and optimizer used.
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:253]:
```suggestion
        Creates a new model.

        :param x_train: Samples of train data.
        :param y_train: Labels of train data.
        :param x_test: Samples of test data.
        :param y_test: Labels of test data.
        :param num_classes: Number of classes of labels in train data.
        :param batch_size: The size of batch used for training.
        :param epochs: The number of epochs for which training need to be applied.
        :return model, loss_fn, optimizer - trained model, loss function used to train the model and optimizer used.
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+                print("Epoch {} train accuracy: {}".format(epoch, train_accuracy))  # pylint: disable=C0209
+        test_accuracy = self.test_accuracy(model, dataloader_test)
+        print("Final test accuracy: {}".format(test_accuracy))  # pylint: disable=C0209
+        return model, loss_fn, optimizer
+
+    @classmethod
+    def test_accuracy(cls, model, test_loader):
+        """
+        Calculates test accuracy on trained model
+        :param model: Trained model.
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:306]:
```suggestion
        Calculates test accuracy on trained model

        :param model: Trained model.
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+        if model_was_training:
+            model.train()
+        return accuracy
+
+    # This function is responsible for returning indices of poison images with maximum gradient norm
+    @classmethod
+    def select_poison_indices(cls, classifier, x_samples, y_samples, num_poison):
+        """
+        Select indices of poisoned samples
+        :classifier: Substitute Model.
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:337]:
```suggestion
        Select indices of poisoned samples

        :classifier: Substitute Model.
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+        indices = indices[-num_poison:]
+        return indices  # this will get only indices for target class
+
+    # This function is responsible for applying trigger patches to the images
+    # fixed - where the trigger is applied at the bottom right of the image
+    # random - where the trigger is applied at random location of the image
+    def apply_trigger_patch(self, x_trigger):
+        """
+        Select indices of poisoned samples
+        :x_trigger: Samples to be used for trigger.
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:371]:
```suggestion
        Select indices of poisoned samples

        :x_trigger: Samples to be used for trigger.
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+        x_train[best_indices_poison] = best_x_poisoned
+        return x_train, y_train
+
+    def get_poison_indices(self):
+        """
+        :return: indices of best poison index
+        """
+        return self.indices_poison
+
+    def model_retraining(self, poisoned_samples):
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:199]: Is this method also used for TensorFlow in a future release? If yes, should we for now raise a warning of the model is in TensorFlow?

File: notebooks/poisoning_attack_sleeper_agent_pytorch.ipynb [hunk 12]
```diff
@@ -0,0 +1,2150 @@
+{
```
Comment (beat-buesser) [inline notebooks/poisoning_attack_sleeper_agent_pytorch.ipynb:1]: Please update `notebooks/README.md` with this new notebook.

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 4 above)
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:136]: I think we should move this exception up to the case checking for TensorFlow v2.

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+
+    def create_model(
+        self,
+        x_train: np.ndarray,
+        y_train: np.ndarray,
+        x_test=None,
+        y_test=None,
+        num_classes=10,
+        batch_size=128,
+        epochs=80,
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:242]: Please add typing.

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 11 above)
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:199]: Please add typing.

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+                running_loss += loss.item()
+            if (epoch % 5 == 0) or epoch == (epochs - 1):
+                train_accuracy = 100 * accuracy / total
+                print("Epoch {} train accuracy: {}".format(epoch, train_accuracy))  # pylint: disable=C0209
+        test_accuracy = self.test_accuracy(model, dataloader_test)
+        print("Final test accuracy: {}".format(test_accuracy))  # pylint: disable=C0209
+        return model, loss_fn, optimizer
+
+    @classmethod
+    def test_accuracy(cls, model, test_loader):
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:303]: Please add typing.

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+
+        # compute the accuracy over all test images
+        accuracy = 100 * accuracy / total
+        if model_was_training:
+            model.train()
+        return accuracy
+
+    # This function is responsible for returning indices of poison images with maximum gradient norm
+    @classmethod
+    def select_poison_indices(cls, classifier, x_samples, y_samples, num_poison):
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:334]: Please add typing.

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+            grad_norms.append(grad_norm.sqrt())
+
+        indices = sorted(range(len(grad_norms)), key=lambda k: grad_norms[k])
+        indices = indices[-num_poison:]
+        return indices  # this will get only indices for target class
+
+    # This function is responsible for applying trigger patches to the images
+    # fixed - where the trigger is applied at the bottom right of the image
+    # random - where the trigger is applied at random location of the image
+    def apply_trigger_patch(self, x_trigger):
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:368]: Please add typing.

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @monshri Thank you very much for your pull request! I have left a few review comments, let me know what you think!

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 6 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:217]: So, you mean we pass some variable - dataset_name='cifar10' in this function? I was trying to pass the data from the notebook - but was getting memory error - so is doing it here

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 11 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:201]: Sure

File: notebooks/poisoning_attack_sleeper_agent_pytorch.ipynb (hunk 12 above)
Comment (monshri) [reply notebooks/poisoning_attack_sleeper_agent_pytorch.ipynb:1]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 4 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py]: Agreed

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 13 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:245]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 11 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:201]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 14 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:307]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 15 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:339]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 16 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:374]: Sure

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 6 above)
Comment (beat-buesser) [reply art/attacks/poisoning/sleeper_agent_attack.py:217]: Yes, passing the data similar as for method `poison`.

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 16 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:374]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 15 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:339]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 11 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:201]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 13 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:245]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 4 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 3 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:66]: Done

File: notebooks/poisoning_attack_sleeper_agent_pytorch.ipynb (hunk 12 above)
Comment (monshri) [reply notebooks/poisoning_attack_sleeper_agent_pytorch.ipynb:1]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 11 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+            grad_norms.append(grad_norm.sqrt())  # type: ignore
+
+        indices = sorted(range(len(grad_norms)), key=lambda k: grad_norms[k])
+        indices = indices[-num_poison:]
+        return indices  # this will get only indices for target class
+
+    # This function is responsible for applying trigger patches to the images
+    # fixed - where the trigger is applied at the bottom right of the image
+    # random - where the trigger is applied at random location of the image
+    def apply_trigger_patch(self, x_trigger: np.ndarray) -> Tuple[Any]:
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:406]:
```suggestion
    def apply_trigger_patch(self, x_trigger: np.ndarray) -> Union[np.ndarray, "torch.Tensor"]:
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+            for x in x_trigger:
+                x_cord = random.randrange(0, x.shape[1] - self.patch.shape[1] + 1)
+                y_cord = random.randrange(0, x.shape[2] - self.patch.shape[2] + 1)
+                x[x_cord : x_cord + patch_size, y_cord : y_cord + patch_size, :] = self.patch
+        if isinstance(self.substitute_classifier, PyTorchClassifier):
+            import torch
+
+            return torch.tensor(np.transpose(x_trigger, [0, 3, 1, 2]))  # type: ignore
+
+        return np.transpose(x_trigger, [0, 3, 1, 2])  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:428]: The transpose should only be necessary of if `self.estimator.channels_first == True`?

Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:428]:
```suggestion
            return torch.tensor(np.transpose(x_trigger, [0, 3, 1, 2]))

        return np.transpose(x_trigger, [0, 3, 1, 2])
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 19]
```diff
@@ -0,0 +1,428 @@
...
+        differentiable_params = [p for p in classifier.model.parameters() if p.requires_grad]
+        for x, y in zip(x_samples, y_samples):
+            image = torch.tensor(x).to(device)
+            label = torch.tensor(y).to(device)
+            loss = criterion(model(image.unsqueeze(0)), label.unsqueeze(0))
+            gradients = torch.autograd.grad(loss, differentiable_params, only_inputs=True)
+            grad_norm = 0
+            for grad in gradients:
+                grad_norm += grad.detach().pow(2).sum()  # type: ignore
+            grad_norms.append(grad_norm.sqrt())  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:397]:
```suggestion
                grad_norm += grad.detach().pow(2).sum()
            grad_norms.append(grad_norm.sqrt())
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 20]
```diff
@@ -0,0 +1,428 @@
...
+    # pylint: disable=W0221
+    def poison(  # type: ignore
+        self,
+        x_trigger: np.ndarray,
+        y_trigger: np.ndarray,
+        x_train: np.ndarray,
+        y_train: np.ndarray,
+        x_test: np.ndarray,
+        y_test: np.ndarray,
+    ) -> Tuple[np.ndarray, np.ndarray]:  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:126]:
```suggestion
    ) -> Tuple[np.ndarray, np.ndarray]:

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 21]
```diff
@@ -0,0 +1,428 @@
...
+            finish_poisoning = self._finish_poison_pytorch
+            initializer = self._initialize_poison_pytorch
+            x_train_target_samples = torch.tensor(
+                np.transpose(x_train_target_samples, [0, 3, 1, 2]), dtype=torch.float32
+            )  # type: ignore
+        else:
+            raise NotImplementedError("SleeperAgentAttack is currently implemented only for PyTorch.")
+
+        # Choose samples to poison.
+        x_trigger = self.apply_trigger_patch(x_trigger)  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:155]:
```suggestion
        x_trigger = self.apply_trigger_patch(x_trigger) 
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 22]
```diff
@@ -0,0 +1,428 @@
...
+            finish_poisoning()
+            B_ = np.mean(B_)  # Averaging B losses from multiple batches.  # pylint: disable=C0103
+            if B_ < best_B:
+                best_B = B_  # pylint: disable=C0103
+                best_x_poisoned = x_poisoned
+                best_indices_poison = self.indices_poison
+
+        if self.verbose > 0:
+            print("Best B-score:", best_B)
+        x_train[self.indices_target[best_indices_poison]] = np.transpose(best_x_poisoned, [0, 2, 3, 1])  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:205]:
```suggestion
        x_train[self.indices_target[best_indices_poison]] = np.transpose(best_x_poisoned, [0, 2, 3, 1])
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 23]
```diff
@@ -0,0 +1,428 @@
...
+        index_target = np.where(y_train.argmax(axis=1) == self.class_target)[0]
+        x_train_target_samples = x_train_samples[index_target]
+        y_train_target_samples = y_train[index_target]
+        return x_train_target_samples, y_train_target_samples
+
+    def get_poison_indices(self) -> List[int]:
+        """
+        :return: indices of best poison index
+        """
+        return self.indices_poison  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:226]:
```suggestion
        return self.indices_poison
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 24]
```diff
@@ -0,0 +1,428 @@
...
+        :param y_train: labels for training data.
+        :param x_test: clean test data.
+        :param y_test: labels for test data.
+        """
+        from art.estimators.classification.pytorch import PyTorchClassifier
+
+        x_train = np.transpose(x_train, [0, 3, 1, 2])
+
+        poisoned_samples = np.asarray(poisoned_samples)
+        x_train[self.indices_target[self.indices_poison]] = poisoned_samples  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:250]:
```suggestion
        x_train[self.indices_target[self.indices_poison]] = poisoned_samples
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 25]
```diff
@@ -0,0 +1,428 @@
...
+        :param x_train: Samples of train data.
+        :param y_train: Labels of train data.
+        :param x_test: Samples of test data.
+        :param y_test: Labels of test data.
+        :param num_classes: Number of classes of labels in train data.
+        :param batch_size: The size of batch used for training.
+        :param epochs: The number of epochs for which training need to be applied.
+        :return model, loss_fn, optimizer - trained model, loss function used to train the model and optimizer used.
+        """
+        import torch
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:289]:
```suggestion
        import torch  # lgtm [py/repeated-import]
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 26]
```diff
@@ -0,0 +1,428 @@
...
+        """
+        Select indices of poisoned samples
+
+        :classifier: Substitute Model.
+        :x_samples: Samples of poison.
+        :y_samples: Labels of samples of poison.
+        :num_poison: Number of poisoned samples to be selected out of all x_samples.
+        :return indices - Indices of samples to be poisoned.
+        """
+        import torch
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:381]:
```suggestion
        import torch  # lgtm [py/repeated-import]
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 27]
```diff
@@ -0,0 +1,428 @@
...
+# SOFTWARE.
+"""
+This module implements Sleeper Agent attack on Neural Networks.
+
+| Paper link: https://arxiv.org/abs/2106.08970
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Any, Tuple, TYPE_CHECKING, List
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:26]:
```suggestion
from typing import Any, Tuple, TYPE_CHECKING, Union, List
```

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 28]
```diff
@@ -0,0 +1,428 @@
...
+        retraining_factor: int = 1,
+        model_retrain: bool = False,
+        model_retraining_epoch: int = 1,
+        class_source: int = 0,
+        class_target: int = 1,
+    ):
+        """
+        Initialize a Sleeper Agent poisoning attack.
+
+        :param classifier: The proxy classifier used for the attack.
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:74]: Pleas order the param descriptions according to arguments in method.

File: art/attacks/poisoning/sleeper_agent_attack.py [hunk 29]
```diff
@@ -0,0 +1,428 @@
...
+        clip_values: Tuple[float, float] = (0, 1.0),
+        verbose: int = 1,
+        indices_target: List[int] = None,
+        patching_strategy: str = "random",
+        selection_strategy: str = "random",
+        retraining_factor: int = 1,
+        model_retrain: bool = False,
+        model_retraining_epoch: int = 1,
+        class_source: int = 0,
+        class_target: int = 1,
```
Comment (beat-buesser) [inline art/attacks/poisoning/sleeper_agent_attack.py:69]:
```suggestion
        self,
        classifier: "CLASSIFIER_NEURALNETWORK_TYPE",
        percent_poison: float,
        patch: np.ndarray,
        indices_target: List[int],
        epsilon: float = 0.1,
        max_trials: int = 8,
        max_epochs: int = 250,
        learning_rate_schedule: Tuple[List[float], List[int]] = ([1e-1, 1e-2, 1e-3, 1e-4], [100, 150, 200, 220]),
        batch_size: int = 128,
        clip_values: Tuple[float, float] = (0, 1.0),
        verbose: int = 1,
        patching_strategy: str = "random",
        selection_strategy: str = "random",
        retraining_factor: int = 1,
        model_retrain: bool = False,
        model_retraining_epoch: int = 1,
        class_source: int = 0,
        class_target: int = 1,
```

Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:69]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 28 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:74]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 27 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:26]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 25 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:289]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 24 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:250]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 23 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:226]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 22 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:205]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 21 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:155]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 20 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:126]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 19 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:397]: Done

File: art/attacks/poisoning/sleeper_agent_attack.py (hunk 18 above)
Comment (monshri) [reply art/attacks/poisoning/sleeper_agent_attack.py:428]: Done

**Review by beat-buesser — APPROVED** [review]

**Pull request merged**